crates/astria-sequencer/src/service/mempool.rs

```diff
@@ -108,228 +125,345 @@ impl Service<MempoolRequest> for Mempool {
 /// as well as stateful checks (nonce and balance checks).
 ///
 /// If the tx passes all checks, status code 0 is returned.
-#[allow(clippy::too_many_lines)]
 #[instrument(skip_all)]
 async fn handle_check_tx<S: accounts::StateReadExt + address::StateReadExt + 'static>(
-    req: request::CheckTx,
+    request::CheckTx {
+        tx,
+        kind,
+    }: request::CheckTx,
     state: S,
-    mempool: &mut AppMempool,
+    mempool: AppMempool,
+    cached_immutable_checks: Arc<Cache<[u8; 32], ImmutableChecksResult>>,
     metrics: &'static Metrics,
 ) -> response::CheckTx {
-    use sha2::Digest as _;
-
-    let start_parsing = Instant::now();
+    let start = Instant::now();
 
-    let request::CheckTx {
-        tx, ..
-    } = req;
-
-    let tx_hash = sha2::Sha256::digest(&tx).into();
+    // So we don't waste time hashing a large object, we don't check the cache before the size
+    // check.
     let tx_len = tx.len();
-
     if tx_len > MAX_TX_SIZE {
         metrics.increment_check_tx_removed_too_large();
-        return response::CheckTx {
-            code: Code::Err(AbciErrorCode::TRANSACTION_TOO_LARGE.value()),
-            log: format!(
-                "transaction size too large; allowed: {MAX_TX_SIZE} bytes, got {}",
-                tx.len()
+        return new_error_response(
+            AbciErrorCode::TRANSACTION_TOO_LARGE,
+            format!(
+                "transaction size too large; allowed: {MAX_TX_SIZE} bytes, got: {tx_len} bytes",
             ),
-            info: AbciErrorCode::TRANSACTION_TOO_LARGE.info(),
-            ..response::CheckTx::default()
-        };
+        );
     }
 
-    let raw_signed_tx = match raw::SignedTransaction::decode(tx) {
+    // Ok to hash the tx now and check in the cache.
+    let tx_hash = sha2::Sha256::digest(&tx).into();
+    let signed_tx = match cached_immutable_checks
+        .get_value_or_guard_async(&tx_hash)
+        .await
+    {
+        Ok(Ok(cached_tx)) => {
```

Comment:
is there a reason for caching the tx instead of just its hash? since you're sent the tx again in `CheckTx`, this seems unnecessary

Reply:
It's to avoid parsing the protobuf tx from bytes and then converting the protobuf tx to a `SignedTransaction`.  It's mainly the second part which is worth avoiding, since we're doing signature checking in `SignedTransaction::try_from_raw`.